src/bootstrap/doc.rs

```diff
@@ -491,37 +471,115 @@ impl Step for Std {
+            if STD_PUBLIC_CRATES.iter().any(|k| *k == requested_crate.as_str()) {
                 let index = out.join(requested_crate).join("index.html");
                 open(builder, &index);
             }
         }
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct JsonStd {
+    pub stage: u32,
+    pub target: TargetSelection,
+}
+
+impl Step for JsonStd {
+    type Output = ();
+    const DEFAULT: bool = false;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.alias("rust-json-docs")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(Std { stage: run.builder.top_stage, target: run.target });
+    }
+
+    /// Build JSON documentation for the standard library crates.
+    ///
+    /// This is largely just a wrapper around `cargo doc`.
+    fn run(self, builder: &Builder<'_>) {
+        let stage = self.stage;
+        let target = self.target;
+        builder.info(&format!("Building JSON documentation for stage{} std ({})", stage, target));
+        let out = builder.json_doc_out(target);
+
+        let extra_args = [
+            OsStr::new("-Z"),
+            OsStr::new("unstable-options"),
+            OsStr::new("--output-format"),
+            OsStr::new("json"),
+            OsStr::new("--resource-suffix"),
+            OsStr::new(&builder.version),
+        ];
+
+        doc_std(builder, stage, target, &out, &extra_args, &[])
+    }
+}
+
+/// Name of the crates that are visible to consumers of the standard library.
+/// Documentation for internal crates is handled by the rustc step, so internal crates will show
+/// up there.
+///
+/// Order here is important!
+/// Crates need to be processed starting from the leaves, otherwise rustdoc will not
+/// create correct links between crates because rustdoc depends on the
+/// existence of the output directories to know if it should be a local
+/// or remote link.
+const STD_PUBLIC_CRATES: [&str; 5] = ["core", "alloc", "std", "proc_macro", "test"];
+
+/// Build the documentation for public standard library crates.
+///
+/// `requested_crates` can be used to build only a subset of the crates. If empty, all crates will
+/// be built.
+fn doc_std(
+    builder: &Builder<'_>,
+    stage: u32,
+    target: TargetSelection,
+    out: &Path,
+    extra_args: &[&OsStr],
+    requested_crates: &[String],
+) {
+    if builder.no_std(target) == Some(true) {
+        panic!(
+            "building std documentation for no_std target {target} is not supported\n\
+             Set `docs = false` in the config to disable documentation."
+        );
+    }
+    let compiler = builder.compiler(stage, builder.config.build);
+    // This is directory where the compiler will place the output of the command.
+    // We will then copy the files from this directory into the final `out` directory, the specified
+    // as a function parameter.
+    let out_dir = builder.stage_out(compiler, Mode::Std).join(target.triple).join("doc");
```

Comment:
This makes me nervous. Say someone runs
```
x doc std
x doc rust-json-docs
```
Then we'll
1. create the HTML docs in `build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc`
2. copy them to `build/x86_64-unknown-linux-gnu/doc`
3. create the JSON docs in `build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc`
4. copy everything in that folder to `build/x86_64-unknown-linux-gnu/json-doc`

So json-doc will end up having both the HTML and the JSON, which is not correct.

I think instead you should introduce an `enum DocMode { HTML, JSON }` and pass that in to this function, then determine `out_dir` and `out` based on that to make sure they're never the same.

Reply:
I'm not sure why this would cause an error but not your previous code. I'm not suggesting that you add any new additional arguments, just that you have a different `out_dir` variable here based on the `DocumentationFormat`.

Reply:
Just to confirm my understanding, this is how I _think_ it works.

`out_dir` is where we retrieve the files that we then copy into the final location.
Just changing `out_dir` without changing the `cargo rustdoc` command is not going to work: `cargo rustdoc` places the JSON docs into the `doc` folder by default; if we change `out_dir` without changing the invocation, we will get a file not found/folder not found error.
That's why I tried to pass the `--out-dir` parameter, to instruct `cargo rustdoc` to store the output files in a completely different (and isolated) folder.

Reply:
Ok, I see the problem, cargo doesn't allow you to configure the target folder structure - `out_dir` is currently an *output* we hard code, not an input.

To avoid messing with this you can instead delete the doc/ directory before running cargo, that should avoid issues where we copy the output into two different dist components.

Reply:
Gotcha, done in [88f96de](https://github.com/rust-lang/rust/pull/101799/commits/88f96de0600f90797d203b27e613e8f2e4b85dec)